gravitee-rest-api-service/src/main/java/io/gravitee/rest/api/service/spring/GraviteeJavaMailSenderImpl.java

```diff
@@ -46,61 +48,85 @@ public GraviteeJavaMailSenderImpl(ParameterService parameterService, EventManage
 
     @Override
     public synchronized Session getSession() {
-        if (!initialized) {
-            this.setHost(parameterService.find(Key.EMAIL_HOST));
-            String port = parameterService.find(Key.EMAIL_PORT);
+        ReferenceContext ref = getReferenceContext();
+        JavaMailSenderImpl mailSender = this.getMailSenderByReference(ref);
+        if (mailSender == null) {
+            mailSender = new JavaMailSenderImpl();
+            mailSender.setHost(parameterService.find(Key.EMAIL_HOST, ref.referenceId, ref.referenceType));
+            String port = parameterService.find(Key.EMAIL_PORT, ref.referenceId, ref.referenceType);
             if (StringUtils.isNumeric(port)) {
-                this.setPort(Integer.parseInt(port));
+                mailSender.setPort(Integer.parseInt(port));
             }
-            this.setUsername(parameterService.find(Key.EMAIL_USERNAME));
-            this.setPassword(parameterService.find(Key.EMAIL_PASSWORD));
-            this.setProtocol(parameterService.find(Key.EMAIL_PROTOCOL));
-            this.setJavaMailProperties(loadProperties());
+            mailSender.setUsername(parameterService.find(Key.EMAIL_USERNAME, ref.referenceId, ref.referenceType));
+            mailSender.setPassword(parameterService.find(Key.EMAIL_PASSWORD, ref.referenceId, ref.referenceType));
+            mailSender.setProtocol(parameterService.find(Key.EMAIL_PROTOCOL, ref.referenceId, ref.referenceType));
+            mailSender.setJavaMailProperties(loadProperties(ref.referenceId, ref.referenceType));
 
-            initialized = true;
+            this.mailSenderByReference.put(ref, mailSender);
         }
 
-        return super.getSession();
+        return mailSender.getSession();
+    }
+
+    @NotNull
+    private ReferenceContext getReferenceContext() {
```

Comment:
I would have prefered to use `GraviteeContext.getCurrentContext();` see next comment